src/materials/NEMLStress.C

```diff
@@ -28,143 +31,132 @@ NEMLStress::validParams()
-                        "NEML xml variable value for neml_variable_iname[3]");
-  params.addParam<Real>("neml_variable_value4",
-                        "NEML xml variable value for neml_variable_iname[4]");
-  params.addParam<Real>("neml_variable_value5",
-                        "NEML xml variable value for neml_variable_iname[5]");
-  params.addParam<Real>("neml_variable_value6",
-                        "NEML xml variable value for neml_variable_iname[6]");
-  params.addParam<Real>("neml_variable_value7",
-                        "NEML xml variable value for neml_variable_iname[7]");
-
+                                            "Names of NEML XML name/value pairs");
+  params.addParam<std::vector<Real>>("neml_variable_value",
+                                     "Corresponding NEML XML variable values");
+  for (size_t i = 0; i < _nvars_max; ++i)
+  {
+    auto istr = Moose::stringify(i);
+    params.addParam<Real>("neml_variable_value" + istr,
+                          "NEML XML variable value for neml_variable_iname[" + istr + "]");
+  }
   return params;
 }
 
-NEMLStress::NEMLStress(const InputParameters & parameters) : NEMLStressBase(parameters)
+NEMLStress::NEMLStress(const InputParameters & parameters)
+  : NEMLStressBase(parameters),
+    _neml_variable_iname(getParam<std::vector<std::string>>("neml_variable_iname")),
+    _neml_variable_value(getParam<std::vector<Real>>("neml_variable_value")),
+    _nvars(_neml_variable_iname.size())
 {
-  FileName fname(getParam<FileName>("database"));
-  std::string mname(getParam<std::string>("model"));
-
-  std::string xmlStringForNeml = loadFileIntoString(fname);
-  if (isParamValid("neml_variable_iname"))
-    replaceXmlVariables(xmlStringForNeml);
-
-  _model = neml::parse_string_unique(xmlStringForNeml, mname);
-}
-
-std::string
-NEMLStress::compareVectorsOfStrings(const std::vector<std::string> & strList1,
-                                    const std::vector<std::string> & strList2) const
-{
-  std::string missingNames;
-  for (auto & str1 : strList1)
+  // fetch and check inames and values for {variable} replacement
+  bool value_vector = false;
+  if (!_neml_variable_value.empty())
   {
-    bool found = false;
-    for (auto & str2 : strList2)
-    {
-      if (str1 == str2)
-      {
-        found = true;
-        break;
-      }
-    }
-    if (!found)
-      missingNames += str1 + " ";
+    if (_nvars != _neml_variable_value.size())
+      paramError("neml_variable_value",
+                 "If any values are specified the number must match the number of entries in "
+                 "neml_variable_iname");
+    value_vector = true;
+  }
+  else if (_nvars > _nvars_max)
+    paramError("neml_variable_iname",
+               "NEMLStressVariableInput can only have up to ",
+               _nvars_max,
+               " neml_variable_iname entries");
+
+  // We permit the numbered parameters to override the values vector.
+  _neml_variable_value.resize(_nvars);
```

Comment:
I guess it doesn't really hurt anything, but it kind of seems odd to resize `_neml_variable_value` and go into the loop below in the case when `neml_variable_value` is set.

Reply:
You mean when it is _not_ set? In that case I'm not going _into_ the loop, because the continuation condition in the head of the loop is never met. I never liked having extra conditionals around such loops, they just seem redundant.